hikari/traits.py

```diff
@@ -560,133 +560,6 @@ class GatewayBotAware(
 
     __slots__: typing.Sequence[str] = ()
 
-    async def join(self, until_close: bool = True) -> None:
-        """Wait indefinitely until the application closes.
-
-        This can be placed in a task and cancelled without affecting the
-        application runtime itself. Any exceptions raised by shards will be
-        propagated to here.
-
-        Other Parameters
-        ----------------
-        until_close : builtins.bool
-            Defaults to `builtins.True`. If set, the waiter will stop as soon as
-            a request for shut down is processed. This can allow you to break
-            and begin closing your own resources.
-
-            If `builtins.False`, then this will wait until all shards' tasks
-            have died.
-        """
-        raise NotImplementedError
-
-    def run(
-        self,
-        *,
-        activity: typing.Optional[presences.Activity] = None,
-        afk: bool = False,
-        close_passed_executor: bool = False,
-        idle_since: typing.Optional[datetime.datetime] = None,
-        ignore_session_start_limit: bool = False,
-        large_threshold: int = 250,
-        status: presences.Status = presences.Status.ONLINE,
-        shard_ids: typing.Optional[typing.AbstractSet[int]] = None,
-        shard_count: typing.Optional[int] = None,
-    ) -> None:
```

Comment:
this is a pretty important breaking change which should be in the changelog, also why was this removed?

Reply:
For consistency with `RESTBotAware`. They are both now `Runnable` and how their lifetime in regards to parameters to `start`, `run`, `join` and `close` are left up to implementation detail.

I forgot to add it to the changelog, will do so

Reply:
For consistency with a class where all the run and start args were considered impl specific you removed the generic run and start args from the trait for a class where some of the run and start args were considered a part of the abstract interface since they were discord detail rather than bot impl detail?

Reply:
They do not have to be Discord detail. They are left up to however the developers want to define them. The GatewayBotAware trait just specifies all the components it should have and thats about it. There is no reason to enforce this as abstraction detail here when it comes to how it should run

Reply:
> There is no reason to enforce this as abstraction detail here when it comes to how it should run.

There is? so ppl can run or start a hikari gateway bot (in-memory shard manager) generically while still being able to control what it declares to Discord on startup (which is what most of these args are for).

Also the field `close_passed_executor` make sense there for the run method since these are both `ExecuterAware` (meaning they manage an executor internally) and I'd argue missing that field for `RESTBotAware.run` was a mistake.

Reply:
But they don't need to necessarily manage that in the run method. They could be done other ways. I just don't see the need for them to be enforced like this here

Reply:
> I just don't see the need for them to be enforced like this here

So ppl can manage them generically, like not everyone just locks their stuff to 1 implementation of Hikari's gateway bot and without that you just won't be able to set some of these fields generically and others you'll need to send another message over the gateway on start-up which is kind of geh.

Also ppl can just not implement GatewayBot specifically if these do not apply to them, like I already do cause GatewayBot also enforces having a cache and none of my implementations have caches.

Also you like enforce that Shard has a start and close method even though not all Shard implementations will be "started" like that.

Reply:
In the same way we dont enforce what the parameters to start, join and close a shard or a `RESTBot` should be, we shouldnt be enforcing it for `GatewayBot` either. Our API standard says that they should be runnable, but it stops there.

With the example you gave before of `ExecutorAware`, we never enforce how that executor should be handled, we just type that an ExecutorAware object will have a executor property.

Reply:
> In the same way we dont enforce what the parameters to start, join and close a shard or a RESTBot should be, we shouldnt be enforcing it for GatewayBot either. Our API standard says that they should be runnable, but it stops there.

There's a big distinction here though, all gateway bots will do stuff like declare an activity on startup. Rest bots don't have anything like that which is why it doesn't have any extra run or start arguments.

> With the example you gave before of ExecutorAware, we never enforce how that executor should be handled, 

Wdym, we were enforcing it and now you're removing it.

Reply:
> In the same way we dont enforce what the parameters to start, join and close a shard

Even just having the method on Shard is enforcing behaviour, its the same thing as having optional arguments. A start method like this would make no sense for a Shard impl which just wraps around a shared message queue as an example and yet its still enforced. Start methods also aren't how you'd want to handle this for a gateway bot impl or shard impl which targets trio or anyio and having start methods on the interface like this is exposing asyncio specific async structuring.

Reply:
We have as an API design for Shard to be be Runnable (this includes having a `start`, `join` and `close`). I am aware that you can have shards without those methods, but by our API design, they should. If someone doesn't want to use this API design, they can define and make their own.

But just like we enforce a shard being runnable, we don't enfoce it to have any specific arguments. This is done in Shard, RESTBotAware, and now GatewayBotAware.

I don't think we should enfoce those arguments at a trait level, which is why I  removed them in this pr.

Reply:
> We have as an API design for Shard to be be Runnable (this includes having a start, join and close). I am aware that you can have shards without those methods, but by our API design, they should. If someone doesn't want to use this API design, they can define and make their own.

Ok so that wouldn't be compatible with Hikari at all so you're just saying "don't use Hikari" here and also why does it need to be a part of the interface? is there even any reason for shard to be directly runnable on the abstract interface when this it even breaks compatibility with the standard gateway bot impl let alone custom implementations.

> But just like we enforce a shard being runnable, we don't enfoce it to have any specific arguments. This is done in Shard, RESTBotAware, and now GatewayBotAware.

Also that's just wrong, the only reason no args are enforced for shard.start is because it just straight out doesn't take any args on the standard implementation (since api `shard.start` is just a copy of the standard implementation's start method) and for `RESTBotAware.run` the reason is because I just missed the close executor arg when i implement it.